File: 01-api-pedidos/manual/java/src/test/java/com/treinamento/api/ApiTest.java
package com.treinamento.api;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.MvcResult;

import static org.junit.jupiter.api.Assertions.*;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;

@SpringBootTest
@AutoConfigureMockMvc
class ApiTest {

    @Autowired
    private MockMvc mockMvc;

    @Autowired
    private ObjectMapper objectMapper;

    @Test
    @DisplayName("CRUD de produto: criar, buscar, atualizar e deletar")
    void testCrudProduto() throws Exception {
        // CREATE
        String novoProduto = """
                {
                    "nome": "Produto Teste CRUD",
                    "descricao": "Descricao do produto de teste",
                    "preco": 99.90,
                    "estoque": 10,
                    "categoria": "Testes"
                }
                """;
        MvcResult criarResult = mockMvc.perform(post("/api/produtos")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(novoProduto))
                .andExpect(status().isCreated())
                .andReturn();

        JsonNode criado = objectMapper.readTree(criarResult.getResponse().getContentAsString());
        long id = criado.get("id").asLong();
        assertEquals("Produto Teste CRUD", criado.get("nome").asText());
        assertEquals(99.90, criado.get("preco").asDouble(), 0.001);

        // READ
        mockMvc.perform(get("/api/produtos/" + id))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.nome").value("Produto Teste CRUD"));

        // UPDATE
        String atualizacao = """
                {
                    "nome": "Produto Atualizado",
                    "preco": 149.90
                }
                """;
        mockMvc.perform(put("/api/produtos/" + id)
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(atualizacao))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.nome").value("Produto Atualizado"))
                .andExpect(jsonPath("$.preco").value(149.90));

        // DELETE
        mockMvc.perform(delete("/api/produtos/" + id))
                .andExpect(status().isNoContent());

        // VERIFY DELETED
        mockMvc.perform(get("/api/produtos/" + id))
                .andExpect(status().isNotFound());
    }

    @Test
    @DisplayName("Busca case-insensitive: 'notebook' deve encontrar 'Notebook Pro'")
    void testBuscaCaseInsensitive() throws Exception {
        MvcResult result = mockMvc.perform(get("/api/produtos")
                        .param("busca", "notebook")
                        .param("page", "1")
                        .param("limit", "20"))
                .andExpect(status().isOk())
                .andReturn();

        String body = result.getResponse().getContentAsString();
        JsonNode json = objectMapper.readTree(body);
        JsonNode data = json.get("data");

        boolean encontrou = false;
        for (JsonNode produto : data) {
            if ("Notebook Pro".equals(produto.get("nome").asText())) {
                encontrou = true;
                break;
            }
        }

        assertTrue(encontrou, "Busca por 'notebook' deveria retornar 'Notebook Pro'");
    }

    @Test
    @DisplayName("Criar pedido: deve criar com sucesso e retornar status 201")
    void testCriarPedido() throws Exception {
        String pedidoBody = """
                {
                    "cliente_id": 1,
                    "itens": [
                        {"produto_id": 1, "quantidade": 2},
                        {"produto_id": 2, "quantidade": 1}
                    ]
                }
                """;

        MvcResult result = mockMvc.perform(post("/api/pedidos")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(pedidoBody))
                .andExpect(status().isCreated())
                .andReturn();

        JsonNode json = objectMapper.readTree(result.getResponse().getContentAsString());
        assertNotNull(json.get("id"));
        assertEquals(1, json.get("cliente_id").asInt());
        assertEquals("pendente", json.get("status").asText());
        assertTrue(json.get("total").asDouble() > 0);

        // Verify details via GET
        long pedidoId = json.get("id").asLong();
        MvcResult detalheResult = mockMvc.perform(get("/api/pedidos/" + pedidoId))
                .andExpect(status().isOk())
                .andReturn();

        JsonNode detalhe = objectMapper.readTree(detalheResult.getResponse().getContentAsString());
        assertEquals(pedidoId, detalhe.get("id").asLong());
        assertNotNull(detalhe.get("itens"));
        assertNotNull(detalhe.get("cliente_nome"));
    }

    @Test
    @DisplayName("Preco negativo deve ser rejeitado com status 400")
    void testPrecoNegativoRejeitado() throws Exception {
        String body = """
                {
                    "nome": "Produto Preco Negativo",
                    "preco": -10.00,
                    "estoque": 5,
                    "categoria": "Teste"
                }
                """;

        mockMvc.perform(post("/api/produtos")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(body))
                .andExpect(status().isBadRequest());
    }

    @Test
    @DisplayName("Listar clientes: deve retornar lista com clientes cadastrados")
    void testListarClientes() throws Exception {
        MvcResult result = mockMvc.perform(get("/api/clientes"))
                .andExpect(status().isOk())
                .andReturn();

        String body = result.getResponse().getContentAsString();
        JsonNode json = objectMapper.readTree(body);

        assertTrue(json.isArray(), "Resposta deve ser um array");
        assertEquals(20, json.size(), "Deve haver 20 clientes no seed");

        // Detail with order history
        MvcResult detalheResult = mockMvc.perform(get("/api/clientes/1"))
                .andExpect(status().isOk())
                .andReturn();

        JsonNode cliente = objectMapper.readTree(detalheResult.getResponse().getContentAsString());
        assertNotNull(cliente.get("pedidos"), "Cliente deve ter campo pedidos");
        assertNotNull(cliente.get("total_pedidos"), "Cliente deve ter campo total_pedidos");
        assertNotNull(cliente.get("total_gasto"), "Cliente deve ter campo total_gasto");
    }

    @Test
    @DisplayName("Paginacao: totalPages deve usar ceil (ex: 15 items / 10 = 2 paginas)")
    void testPaginacaoTotalPages() throws Exception {
        MvcResult result = mockMvc.perform(get("/api/produtos")
                        .param("page", "1")
                        .param("limit", "3"))
                .andExpect(status().isOk())
                .andReturn();

        String body = result.getResponse().getContentAsString();
        JsonNode json = objectMapper.readTree(body);

        long total = json.get("total").asLong();
        int limit = json.get("limit").asInt();
        int totalPages = json.get("total_pages").asInt();

        int expectedTotalPages = (int) Math.ceil((double) total / limit);
        assertEquals(expectedTotalPages, totalPages,
                "totalPages deveria ser " + expectedTotalPages + " mas foi " + totalPages);
    }

    @Test
    @DisplayName("Dashboard resumo: deve retornar totalVendas, totalPedidos e ticketMedio")
    void testDashboardResumo() throws Exception {
        MvcResult result = mockMvc.perform(get("/api/dashboard/resumo"))
                .andExpect(status().isOk())
                .andReturn();

        JsonNode json = objectMapper.readTree(result.getResponse().getContentAsString());

        assertNotNull(json.get("total_vendas"), "Campo total_vendas deve existir");
        assertNotNull(json.get("total_pedidos"), "Campo total_pedidos deve existir");
        assertNotNull(json.get("ticket_medio"), "Campo ticket_medio deve existir");

        assertTrue(json.get("total_vendas").asDouble() > 0, "total_vendas deve ser maior que zero");
        assertTrue(json.get("total_pedidos").asLong() > 0, "total_pedidos deve ser maior que zero");
        assertTrue(json.get("ticket_medio").asDouble() > 0, "ticket_medio deve ser maior que zero");
    }

    @Test
    @DisplayName("Total do pedido deve considerar quantidade: 3 unidades x R$100 = R$300")
    void testTotalPedidoComQuantidade() throws Exception {
        String novoProduto = """
                {
                    "nome": "Produto Para Teste Quantidade",
                    "descricao": "Produto de teste",
                    "preco": 100.00,
                    "estoque": 50,
                    "categoria": "Teste"
                }
                """;
        MvcResult prodResult = mockMvc.perform(post("/api/produtos")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(novoProduto))
                .andExpect(status().isCreated())
                .andReturn();

        JsonNode prodJson = objectMapper.readTree(prodResult.getResponse().getContentAsString());
        long produtoId = prodJson.get("id").asLong();

        String pedidoBody = String.format("""
                {
                    "cliente_id": 1,
                    "itens": [
                        {"produto_id": %d, "quantidade": 3}
                    ]
                }
                """, produtoId);

        MvcResult pedidoResult = mockMvc.perform(post("/api/pedidos")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(pedidoBody))
                .andExpect(status().isCreated())
                .andReturn();

        JsonNode pedidoJson = objectMapper.readTree(pedidoResult.getResponse().getContentAsString());
        double total = pedidoJson.get("total").asDouble();

        assertEquals(300.00, total, 0.001,
                "Total deveria ser 300.00 (3 x 100.00), mas foi " + total);
    }

    @Test
    @DisplayName("Filtro por status: GET /api/pedidos?status=entregue deve retornar apenas pedidos entregues")
    void testFiltroStatus() throws Exception {
        MvcResult allResult = mockMvc.perform(get("/api/pedidos")
                        .param("limit", "200"))
                .andExpect(status().isOk())
                .andReturn();

        JsonNode allJson = objectMapper.readTree(allResult.getResponse().getContentAsString());
        long totalGeral = allJson.get("total").asLong();

        MvcResult result = mockMvc.perform(get("/api/pedidos")
                        .param("status", "entregue")
                        .param("limit", "200"))
                .andExpect(status().isOk())
                .andReturn();

        String body = result.getResponse().getContentAsString();
        JsonNode json = objectMapper.readTree(body);
        JsonNode data = json.get("data");

        for (JsonNode pedido : data) {
            assertEquals("entregue", pedido.get("status").asText(),
                    "Filtro por status=entregue retornou pedido com status diferente");
        }

        assertTrue(json.get("total").asLong() < totalGeral,
                "Filtro por status deveria retornar menos pedidos que o total");
    }
}

File: 01-api-pedidos/ai/php/tests/ApiTest.php
<?php

declare(strict_types=1);

namespace Tests;

use App\Controllers\ClienteController;
use App\Controllers\DashboardController;
use App\Controllers\PedidoController;
use App\Controllers\ProdutoController;
use App\Database;
use App\Middleware\CorsMiddleware;
use PDO;
use PHPUnit\Framework\TestCase;
use Slim\Factory\AppFactory;
use Slim\Psr7\Factory\ServerRequestFactory;
use Slim\Psr7\Factory\StreamFactory;

class ApiTest extends TestCase
{
    private static PDO $db;
    private \Slim\App $app;

    public static function setUpBeforeClass(): void
    {
        self::$db = Database::createTestInstance();
    }

    protected function setUp(): void
    {
        $db = self::$db;

        $app = AppFactory::create();
        $app->addBodyParsingMiddleware();
        $app->addRoutingMiddleware();
        $app->addErrorMiddleware(false, false, false);
        $app->add(CorsMiddleware::class);

        $app->options('/{routes:.+}', function ($request, $response) {
            return $response;
        });

        // Produtos
        $app->get('/api/produtos', function ($req, $res) use ($db) {
            return (new ProdutoController($db))->index($req, $res);
        });
        $app->get('/api/produtos/{id:[0-9]+}', function ($req, $res, $args) use ($db) {
            return (new ProdutoController($db))->show($req, $res, $args);
        });
        $app->post('/api/produtos', function ($req, $res) use ($db) {
            return (new ProdutoController($db))->create($req, $res);
        });
        $app->put('/api/produtos/{id:[0-9]+}', function ($req, $res, $args) use ($db) {
            return (new ProdutoController($db))->update($req, $res, $args);
        });
        $app->delete('/api/produtos/{id:[0-9]+}', function ($req, $res, $args) use ($db) {
            return (new ProdutoController($db))->delete($req, $res, $args);
        });

        // Pedidos
        $app->get('/api/pedidos', function ($req, $res) use ($db) {
            return (new PedidoController($db))->index($req, $res);
        });
        $app->get('/api/pedidos/{id:[0-9]+}', function ($req, $res, $args) use ($db) {
            return (new PedidoController($db))->show($req, $res, $args);
        });
        $app->post('/api/pedidos', function ($req, $res) use ($db) {
            return (new PedidoController($db))->create($req, $res);
        });
        $app->patch('/api/pedidos/{id:[0-9]+}/status', function ($req, $res, $args) use ($db) {
            return (new PedidoController($db))->updateStatus($req, $res, $args);
        });

        // Clientes
        $app->get('/api/clientes', function ($req, $res) use ($db) {
            return (new ClienteController($db))->index($req, $res);
        });
        $app->get('/api/clientes/{id:[0-9]+}', function ($req, $res, $args) use ($db) {
            return (new ClienteController($db))->show($req, $res, $args);
        });

        // Dashboard
        $app->get('/api/dashboard/resumo', function ($req, $res) use ($db) {
            return (new DashboardController($db))->resumo($req, $res);
        });

        $this->app = $app;
    }

    private function request(string $method, string $uri, ?array $body = null): array
    {
        $factory = new ServerRequestFactory();
        $request = $factory->createServerRequest($method, $uri);

        if ($body !== null) {
            $streamFactory = new StreamFactory();
            $json = json_encode($body);
            $stream = $streamFactory->createStream($json);
            $request = $request
                ->withHeader('Content-Type', 'application/json')
                ->withBody($stream)
                ->withParsedBody($body);
        }

        $response = $this->app->handle($request);
        $status = $response->getStatusCode();
        $bodyStr = (string)$response->getBody();
        $data = json_decode($bodyStr, true);

        return [$status, $data];
    }

    public function testCrudProduto(): void
    {
        // CREATE
        [$status, $data] = $this->request('POST', '/api/produtos', [
            'nome'      => 'Produto Teste CRUD',
            'descricao' => 'Descricao do produto de teste',
            'preco'     => 99.90,
            'estoque'   => 10,
            'categoria' => 'Teste',
        ]);
        $this->assertSame(201, $status, 'POST /api/produtos deve retornar 201');
        $this->assertArrayHasKey('id', $data);
        $this->assertSame('Produto Teste CRUD', $data['nome']);
        $this->assertSame(99.90, (float)$data['preco']);
        $id = (int)$data['id'];

        // READ
        [$status, $data] = $this->request('GET', "/api/produtos/{$id}");
        $this->assertSame(200, $status);
        $this->assertSame('Produto Teste CRUD', $data['nome']);

        // UPDATE
        [$status, $data] = $this->request('PUT', "/api/produtos/{$id}", [
            'nome'    => 'Produto Atualizado',
            'preco'   => 149.90,
            'estoque' => 5,
        ]);
        $this->assertSame(200, $status);
        $this->assertSame('Produto Atualizado', $data['nome']);
        $this->assertSame(149.90, (float)$data['preco']);

        // DELETE
        [$status, $data] = $this->request('DELETE', "/api/produtos/{$id}");
        $this->assertSame(204, $status);

        // VERIFY DELETED
        [$status,] = $this->request('GET', "/api/produtos/{$id}");
        $this->assertSame(404, $status);
    }

    public function testBuscaCaseInsensitive(): void
    {
        [$status, $data] = $this->request('GET', '/api/produtos?busca=notebook');

        $this->assertSame(200, $status);
        $this->assertGreaterThan(
            0,
            $data['total'],
            'Busca por "notebook" deveria encontrar "Notebook Pro"'
        );
    }

    public function testCriarPedido(): void
    {
        [$status, $data] = $this->request('POST', '/api/pedidos', [
            'cliente_id' => 1,
            'itens'      => [
                ['produto_id' => 1, 'quantidade' => 2],
                ['produto_id' => 2, 'quantidade' => 1],
            ],
        ]);

        $this->assertSame(201, $status, 'POST /api/pedidos deve retornar 201');
        $this->assertArrayHasKey('id', $data);
        $this->assertSame(1, (int)$data['cliente_id']);
        $this->assertSame('pendente', $data['status']);

        $pedidoId = (int)$data['id'];
        [$status, $pedido] = $this->request('GET', "/api/pedidos/{$pedidoId}");
        $this->assertSame(200, $status);
        $this->assertSame($pedidoId, (int)$pedido['id']);
        $this->assertArrayHasKey('itens', $pedido);
        $this->assertArrayHasKey('cliente_nome', $pedido);
    }

    public function testPrecoNegativoRejeitado(): void
    {
        [$status, $data] = $this->request('POST', '/api/produtos', [
            'nome'  => 'Produto Preco Negativo',
            'preco' => -10.00,
        ]);

        $this->assertSame(
            400,
            $status,
            'Preco negativo deve ser rejeitado com status 400'
        );
    }

    public function testListarClientes(): void
    {
        [$status, $data] = $this->request('GET', '/api/clientes');

        $this->assertSame(200, $status, 'GET /api/clientes deve retornar 200');
        $this->assertIsArray($data);
        $this->assertCount(20, $data, 'Deve haver 20 clientes no seed');

        [$status, $cliente] = $this->request('GET', '/api/clientes/1');
        $this->assertSame(200, $status);
        $this->assertArrayHasKey('pedidos', $cliente);
        $this->assertArrayHasKey('total_pedidos', $cliente);
        $this->assertArrayHasKey('total_gasto', $cliente);
        $this->assertSame('Ana Paula Ferreira', $cliente['nome']);
    }

    public function testPaginacaoTotalPages(): void
    {
        [$status, $data] = $this->request('GET', '/api/produtos?page=1&limit=7');

        $this->assertSame(200, $status);
        $total = (int)$data['total'];
        $limit = (int)$data['limit'];
        $totalPages = (int)$data['totalPages'];

        $expectedPages = (int)ceil($total / $limit);
        $this->assertSame(
            $expectedPages,
            $totalPages,
            "totalPages deve ser ceil({$total}/{$limit})={$expectedPages}, mas foi {$totalPages}"
        );
    }

    public function testDashboardResumo(): void
    {
        [$status, $data] = $this->request('GET', '/api/dashboard/resumo');

        $this->assertSame(200, $status, 'GET /api/dashboard/resumo deve retornar 200');
        $this->assertArrayHasKey('total_vendas', $data);
        $this->assertArrayHasKey('total_pedidos', $data);
        $this->assertArrayHasKey('ticket_medio', $data);

        $this->assertGreaterThan(0, $data['total_vendas'], 'total_vendas deve ser positivo');
        $this->assertGreaterThan(0, $data['total_pedidos'], 'total_pedidos deve ser positivo');
        $this->assertGreaterThan(0, $data['ticket_medio'], 'ticket_medio deve ser positivo');

        $expectedTicket = round($data['total_vendas'] / $data['total_pedidos'], 2);
        $this->assertEqualsWithDelta($expectedTicket, $data['ticket_medio'], 0.01);
    }

    public function testTotalPedidoComQuantidade(): void
    {
        [$status, $produto] = $this->request('POST', '/api/produtos', [
            'nome'      => 'Produto Para Teste Quantidade',
            'descricao' => 'Produto de teste',
            'preco'     => 100.00,
            'estoque'   => 50,
            'categoria' => 'Teste',
        ]);
        $this->assertSame(201, $status);
        $produtoId = (int)$produto['id'];

        [$status, $pedido] = $this->request('POST', '/api/pedidos', [
            'cliente_id' => 1,
            'itens'      => [
                ['produto_id' => $produtoId, 'quantidade' => 3],
            ],
        ]);

        $this->assertSame(201, $status);
        $this->assertEqualsWithDelta(
            300.00,
            (float)$pedido['total'],
            0.01,
            "Total do pedido deveria ser 300.00 (3 x 100.00), mas foi {$pedido['total']}"
        );
    }

    public function testFiltroStatus(): void
    {
        [$status, $todos] = $this->request('GET', '/api/pedidos?limit=200');
        $this->assertSame(200, $status);

        [$status, $entregues] = $this->request('GET', '/api/pedidos?status=entregue&limit=200');
        $this->assertSame(200, $status);

        foreach ($entregues['data'] as $pedido) {
            $this->assertSame(
                'entregue',
                $pedido['status'],
                'Filtro por status=entregue retornou pedido com status diferente'
            );
        }

        $this->assertLessThan(
            (int)$todos['total'],
            (int)$entregues['total'],
            'Filtro por status deveria retornar menos pedidos que o total'
        );
    }
}
